src/librustc/middle/cstore.rs

```diff
@@ -597,56 +532,31 @@ pub mod tls {
     }
 
     /// Execute f after pushing the given DecodingContext onto the TLS stack.
-    pub fn enter_decoding_context<'tcx, F, R>(dcx: &DecodingContext<'tcx>,
-                                              decoder: &mut OpaqueDecoder,
-                                              f: F) -> R
-        where F: FnOnce(&DecodingContext<'tcx>, &mut OpaqueDecoder) -> R
+    pub fn enter_decoding_context<'tcx, F, R>(dcx: &DecodingContext<'tcx>, f: F) -> R
+        where F: FnOnce(&DecodingContext<'tcx>) -> R
     {
-        let tls_payload = (dcx as *const _, decoder as *mut _);
+        let tls_payload = dcx as *const _;
         let tls_ptr = &tls_payload as *const _ as *const TlsPayload;
         TLS_DECODING.with(|tls| {
             let prev = tls.get();
             tls.set(Some(tls_ptr));
-            let ret = f(dcx, decoder);
+            let ret = f(dcx);
             tls.set(prev);
-            return ret
+            ret
         })
     }
 
-    /// Execute f with access to the thread-local decoding context and
-    /// rbml decoder. This function will panic if the decoder passed in and the
-    /// context decoder are not the same.
-    ///
-    /// Note that this method is 'practically' safe due to its checking that the
-    /// decoder passed in is the same as the one in TLS, but it would still be
-    /// possible to construct cases where the DecodingContext is exchanged
-    /// while the same decoder is used, thus working with a wrong context.
-    pub fn with_decoding_context<'decoder, 'tcx, D, F, R>(d: &'decoder mut D, f: F) -> R
-        where D: serialize::Decoder,
-              F: FnOnce(&DecodingContext<'tcx>,
-                        &mut OpaqueDecoder) -> R,
-              'tcx: 'decoder
+    /// Execute f with access to the thread-local decoding context.
+    /// FIXME(eddyb) This is horribly unsound as it allows the
```

Comment:
Let's open a bug to represent the lack of lifetime dispatch testing and tag this FIXME with it.